lockfile/lockfile-types/src/index.ts

```diff
@@ -135,3 +136,24 @@ export type PackageBin = string | { [name: string]: string }
  * }
  */
 export type ResolvedDependencies = Record<string, string>
+
+export interface CatalogSnapshots {
+  [catalogName: string]: { [dependencyName: string]: ResolvedCatalogEntry }
+}
+
+export interface ResolvedCatalogEntry {
+  /**
+   * The real specifier that should be used for this dependency's catalog entry.
+   * This would be the ^1.2.3 portion of:
+   *
+   * @example
+   * catalog:
+   *   foo: ^1.2.3
+   */
+  readonly specifier: string
+
+  /**
+   * The concrete version that the requested specifier resolved to. Ex: 1.2.3
+   */
+  readonly version: string
```

Comment:
In the `pnpm-lock.yaml` file, this will look like:

```yaml
catalogs:
  default:
    foo:
      specifier: ^1.2.3
      version: 1.2.3
```

This PR writes the `version` part. A future PR will read from it to optimize new usages of the catalog protocol. I think the logic will be similar to the existing `replaceVersionInPref` logic.

Reply:
what if there will be multiple named catalogs with the same specs? will the entries be duplicated? Like

```
catalogs:
  default:
    foo:
      specifier: ^1.2.3
      version: 1.2.3
  named:
    foo:
      specifier: ^1.2.3
      version: 1.2.3
```

Reply:
Yup, they'll be duplicated. I think that's okay and somewhat by design if someone duplicates a catalog entry across multiple named catalogs.